### ISSUE TITLE
Sorted topk behaves inconsistently depending on k

### ISSUE THREAD

**Issue by volcacius** [pytorch/pytorch#27542]
## 🐛 Bug

torch.topk with sorted=True doesn't return a result that is consistent across different values of k when dealing with duplicates values. The position of duplicated values in the returned sorted indices varies with k. The behavior varies between CPU and CUDA, and it's inconsistent both between and within the two backends.

This affects the reference implementation for computing accuracy in e.g. imagenet classification ([link](https://github.com/pytorch/examples/blob/ee964a2eeb41e1712fe719b83645c79bcbd0ba1a/imagenet/main.py#L407)), in the sense that passing topk=(1,5) or topk=(1,10) might give different top1 accuracies. The effect is especially notable on highly quantized models, where it's more common to have duplicated values in the output of a layer.

## To Reproduce

On CPU:

```python
In [1]: import torch

In [2]: x = torch.Tensor([1, 2, 5, 4, 5])

In [3]: print(x.topk(1, largest=True, sorted=True))
torch.return_types.topk(
values=tensor([5.]),
indices=tensor([4]))

In [4]: print(x.topk(2, largest=True, sorted=True))
torch.return_types.topk(
values=tensor([5., 5.]),
indices=tensor([2, 4]))

In [5]: print(x.topk(3, largest=True, sorted=True))
torch.return_types.topk(
values=tensor([5., 5., 4.]),
indices=tensor([2, 4, 3]))

In [6]: print(x.topk(4, largest=True, sorted=True))
torch.return_types.topk(
values=tensor([5., 5., 4., 2.]),
indices=tensor([2, 4, 3, 1]))

In [7]: print(x.topk(5, largest=True, sorted=True))
torch.return_types.topk(
values=tensor([5., 5., 4., 2., 1.]),
indices=tensor([2, 4, 3, 1, 0]))
```

On CUDA:

```python
In [1]: import torch

In [2]: y = torch.Tensor([1, 2, 5, 4, 5]).cuda()

In [3]: print(y.topk(1, largest=True, sorted=True))
torch.return_types.topk(
values=tensor([5.], device='cuda:0'),
indices=tensor([2], device='cuda:0'))

In [4]: print(y.topk(2, largest=True, sorted=True))
torch.return_types.topk(
values=tensor([5., 5.], device='cuda:0'),
indices=tensor([4, 2], device='cuda:0'))

In [5]: print(y.topk(3, largest=True, sorted=True))
torch.return_types.topk(
values=tensor([5., 5., 4.], device='cuda:0'),
indices=tensor([4, 2, 3], device='cuda:0'))

In [6]: print(y.topk(4, largest=True, sorted=True))
torch.return_types.topk(
values=tensor([5., 5., 4., 2.], device='cuda:0'),
indices=tensor([2, 4, 3, 1], device='cuda:0'))

In [7]: print(y.topk(5, largest=True, sorted=True))
torch.return_types.topk(
values=tensor([5., 5., 4., 2., 1.], device='cuda:0'),
indices=tensor([2, 4, 3, 1, 0], device='cuda:0'))
```

## Expected behavior

The indices should be always in the same order independently of the value of k. Ideally they should also be in the same order between CPU and GPU. 

## Environment

Reproduced on latest pytorch-nightly.

```
python collect_env.py
Collecting environment information...
PyTorch version: 1.3.0.dev20190917
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 16.04.6 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
CMake version: Could not collect

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration:
GPU 0: Quadro P6000
GPU 1: Quadro P6000
GPU 2: Quadro P6000

Nvidia driver version: 418.67
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.5.0

Versions of relevant libraries:
[pip] numpy==1.17.2
[pip] torch==1.3.0.dev20190917
[pip] torchvision==0.5.0a0+e8b830f
[conda] blas                      1.0                         mkl
[conda] mkl                       2019.4                      243
[conda] mkl-service               2.3.0            py36he904b0f_0
[conda] mkl_fft                   1.0.14           py36ha843d7b_0
[conda] mkl_random                1.1.0            py36hd6b4f25_0
[conda] pytorch                   1.3.0.dev20190917 py3.6_cuda10.0.130_cudnn7.6.2_0    pytorch-nightly
[conda] torchvision               0.5.0.dev20190917      py36_cu100    pytorch-nightly
```

**Comment by cpuhrsch** [comment]
I wrote a script that makes this easier to reproduce and was able to reproduce it as well

```
$ cat /tmp/repor.py
import torch
import sys


def make_xs(x):
    xs = []
    xs.append(x.topk(1, largest=True, sorted=True))
    xs.append(x.topk(2, largest=True, sorted=True))
    xs.append(x.topk(3, largest=True, sorted=True))
    xs.append(x.topk(4, largest=True, sorted=True))
    xs.append(x.topk(5, largest=True, sorted=True))
    return xs

for try_i in range(10):
    xs_cpu = make_xs(torch.Tensor([1, 2, 5, 4, 5]))
    xs_cuda = make_xs(torch.Tensor([1, 2, 5, 4, 5]).cuda())
    if not all((xs_cpu[i].indices == xs_cuda[i].indices.cpu()).all() for i in range(len(xs_cpu))):
        for i in range(len(xs_cpu)):
            if (xs_cpu[i].indices != xs_cuda[i].indices.cpu()).any():
                print("fail on try {}".format(try_i))
                print("cpu:\n" + str(xs_cpu))
                print("cuda:\n" + str(xs_cuda))
                sys.exit(1)

$ python /tmp/repor.py
fail on try 0
cpu:
[torch.return_types.topk(
values=tensor([5.]),
indices=tensor([2])), torch.return_types.topk(
values=tensor([5., 5.]),
indices=tensor([2, 4])), torch.return_types.topk(
values=tensor([5., 5., 4.]),
indices=tensor([2, 4, 3])), torch.return_types.topk(
values=tensor([5., 5., 4., 2.]),
indices=tensor([2, 4, 3, 1])), torch.return_types.topk(
values=tensor([5., 5., 4., 2., 1.]),
indices=tensor([2, 4, 3, 1, 0]))]
cuda:
[torch.return_types.topk(
values=tensor([5.], device='cuda:0'),
indices=tensor([2], device='cuda:0')), torch.return_types.topk(
values=tensor([5., 5.], device='cuda:0'),
indices=tensor([4, 2], device='cuda:0')), torch.return_types.topk(
values=tensor([5., 5., 4.], device='cuda:0'),
indices=tensor([4, 2, 3], device='cuda:0')), torch.return_types.topk(
values=tensor([5., 5., 4., 2.], device='cuda:0'),
indices=tensor([2, 4, 3, 1], device='cuda:0')), torch.return_types.topk(
values=tensor([5., 5., 4., 2., 1.], device='cuda:0'),
indices=tensor([2, 4, 3, 1, 0], device='cuda:0'))]
```

**Comment by volcacius** [comment]
To clarify, the issue is not just that the indices are different between CPU and CUDA. The problem is that the order of duplicate values changes with k. In my CUDA example, for k=1 or 4 or 5 the largest value is reported as being at index 2, while for k=2 or 3 the largest value is reported as being at index 4. The largest value should always be the same, independently of how many k largest values you are asking for.

**Comment by Lalaland** [comment]
@volcacius What do you think the optimal behavior should be for duplicates? What about sorting by the index for duplicates so that duplicates at lower indices are earlier in the list?

**Comment by cpuhrsch** [comment]
I marked this issue for triage review so we can state a choice on the behavior here.

**Comment by volcacius** [comment]
I would check what other frameworks do in this particular context and try to match that. Or it might be a an additional flag that the user can control.
I also would like to report that I'm experiencing inconsistent behavior depending on the batch size of the tensor passed, although I haven't been able to reproduce the problem systematically yet.

**Comment by albanD** [comment]
The `sorted` keyword ensure that they are sorted.
Is the request here to have an option to get a stable sort for the keys?

**Comment by Lalaland** [comment]
It seems like the solution for this is to:

1. Implement a stable implementation of topk that handles ties better.
2. Measure the performance impact of #1 relative to the current implementation.
3. If it's cheap, roll that out as the default and remove the unstable implementation.
    Otherwise, if it's expensive, add a flag for stable behavior.

**Comment by albanD** [comment]
That sounds good. Will you have some time to work on that implementation?

**Comment by gchanan** [comment]
I marked this as "topic: determinism".

I don't see any reason this should be marked high priority.

**Comment by volcacius** [comment]
It affects accuracy computation of quantized image classification models, which is probably the main benchmark scenario when it comes to quantization. With the release of official quantization support in 1.3, you are going to see more users being affected by it. I think it would be a missed opportunity to not prioritize it, integer arithmetic has much more of a chance of being reproducible across systems compared to fp, but this breaks everything at the very last step.

**Comment by Lalaland** [comment]
Feel free to assign this to me. I can quickly try making the CPU version more stable and benchmark that. this weekend. I looked briefly at the GPU implementation of topk, it's a mess though and might need more work.

**Comment by gchanan** [comment]
A bit about the (lack of) high priority designation: I don't think marking a single non-determinism issue really makes sense if we don't make a serious effort at improving determinism globally -- right now we just have a couple of notes about determinism, but it's not straightforward to enforce determinism.  So any one change has limited value.

That doesn't mean it's not worth doing (particularly if there is no perf cost), it just means other things rank higher currently.

**Comment by volcacius** [comment]
I understand what you are saying and I definitely share the sentiment when it comes to numerical computing in general.

My point was really more specific about reproducibility in the context of a (fake) integer datapath. In my experience the relative error introduced by leveraging different and possibly inconsistent floating point backends is often going to be rounded away when you are doing fake quantization, but there are some things that you have to do consistently. Rounding itself is one of them, and it was fixed a couple of versions ago I think, and sorting of your output is definitely another one. 

Just to say that in my opinion, in this particular context, some changes have more value than others. But I also get that there other tasks that are going to be prioritized.

**Comment by gchanan** [comment]
@volcacius:

Yes, I think we are in agreement.  I thought I just owed more explanation than "not high priority".

**Comment by Baranowski** [comment]
Following up on https://github.com/pytorch/pytorch/issues/27542#issuecomment-542833006:

> A bit about the (lack of) high priority designation: I don't think marking a single non-determinism issue really makes sense if we don't make a serious effort at improving determinism globally -- right now we just have a couple of notes about determinism, but it's not straightforward to enforce determinism. So any one change has limited value.
> 
> That doesn't mean it's not worth doing (particularly if there is no perf cost), it just means other things rank higher currently.

Should the `high priority` label be removed?

I was thinking about asking Lalaland whether I may take over but I would do that only if this issue is considered high-pri.

**Comment by albanD** [comment]
Triage review to answer the question above.

**Comment by Lalaland** [comment]
Sorry, I haven't had a decent chance to look at this as other things have come up (and I don't know when I will get an opportunity). My apologies. Feel free to unassign me.

**Comment by mruberry** [comment]
Removing high priority per review.

**Comment by jg-codes** [comment]
Two measurements from KV-cache eviction work (top-k over per-token scores in the cache dtype), torch 2.14.0+cpu on CPU and torch 2.7.1+cu126 on an A100. 8 x 4096 uniform random scores in [0, 1) have 32743 distinct values in float32, 5442 in float16 and 1054 in bfloat16 (same counts on both devices), so in bfloat16 about 97 % of the values share their value with another one. A pure-PyTorch stable selection (order positions by the tie preference, `argsort(descending=True, stable=True)` on the scores in that order, take the first k) costs 1.67-2.66x plain `topk` for shape (1, 8, L), k = L/2, L in {4096, 8192, 32768}, on both devices. Repeated calls and batch sizes 2 to 128 gave identical indices on both devices, so this concerns the unspecified tie order, not kernel nondeterminism.

Since #88227 settled the semantics (lower index wins), the smallest step would be to state in the `torch.topk` documentation that the choice among equal values is unspecified; a `stable=` argument as in `torch.sort` would be the second step.

Drafted with agentic AI assistance; human-reviewed before posting.
